MathLang.hs

```diff
@@ -18,9 +19,12 @@ data Expr = Add
           | Equ
           | If Prog Prog
           | Less
+          | Dup
+          | ExprList [Expr]
+          | IsType
    deriving (Eq, Show)
 
-data Stmt = While Expr Cmd
+data Stmt = While Expr Prog
           | Begin Prog
```

Comment:
What if we changed the `Stmt` data type to the following then drop the `ExprList [Expr]` constructor:
```haskell
data Stmt = While [Expr] Stmt | Begin Prog
```
This way, a list of expressions could be evaluated at the beginning of the while loop (where one condition results in simply a list with only one expression in it) and the following code block would have to be the `Begin Prog` constructor that subsequently executes a contained list of commands.

Reply:
Ooh that sounds like a good idea!

I changed `data Stmt` back to your suggestion, but then I got stuck for 15 minutes trying to implement the `stmt While` valuation function;

```haskell
stmt :: Stmt -> Domain
stmt (While (e:es) c) q fs = case (expr e q fs) of 
                                 (Just ((B True):qs)) -> case (prog c qs fs) of
                                                            Just q -> stmt (While e c) q fs
                                                            _      -> Nothing
                                 (Just (_:qs))        -> Just (qs)
                                 _                    -> Nothing
```

So now `While` can take a list of expressions, but I'm unsure how to evaluate them using recursion.
What I'm trying to do is run `expr` on each `Expr` in `(e:es)`, and then pattern match on the result of that -- the way the above code does -- but the issue I'm having is that `expr` returns a new stack with the result of the evaluation on it, not a list of expressions left to evaluate. That leaves my only reference to the rest of the `Expr`s being `es`, but I can't call `expr` on `es` since `es` is a `[Expr]` and not a `Expr`. 
Sorry, I hope this is making some sense haha, it's pretty late and my brain is jumbled. I saw your review and figured it would be an easy change and it probably is to be honest, I just need to take a look when I've had a little more sleep :) If you had any ideas on how to easily implement this, hit me up, otherwise I'll try again tomorrow